### ISSUE TITLE
`getCode`/`getDeployedCode` wont find a valid artifact

### ISSUE THREAD

**Issue by meetmangukiya** [foundry-rs/foundry#8842]
### Component

Forge

### Have you ensured that all of these are up to date?

- [ ] Foundry
- [ ] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (be451fb 2024-09-11T00:20:08.082212000Z)

### What command(s) is the bug in?

forge test

### Operating System

macOS (Apple Silicon)

### Describe the bug

Fails to find valid artifact if the said contract is not imported in some test file or contract file. But `forge build` finds it and is compiling it fine. See repro repo at https://github.com/meetmangukiya/foundry-get-code-bug-repro

**Comment by meetmangukiya** [comment]
Maybe related/caused by https://github.com/foundry-rs/foundry/pull/7334

**Comment by yash-atreya** [comment]
> Maybe related/caused by #7334

Yes, after #7334, you must import the necessary `.sol` file whose code you're trying to retrieve with `vm.getCode`. The rationale was to speed up compilation times when using filters. This serves as an indicator to the compiler to not ignore that file. 

cc @klkvr

**Comment by klkvr** [comment]
yeah, if you need anything to always be available for `getCode`, you can just put it into src/, it will always get compiled

**Comment by meetmangukiya** [comment]
It's not a relevant source file, its a test fixture(in my usecase which led to me coming across this problem). And importing is not an option because of both files having incompatible compiler pragmas. Maybe worth exploring a config option for declaring test dependencies or if thats complicated, an _always compile_ option?

**Comment by mds1** [comment]
Hm seems like a valid concern so reopening, what do you think @klkvr?

**Comment by klkvr** [comment]
> Hm seems like a valid concern so reopening, what do you think @klkvr?

This is something we've been already considering for a while (allowing to import files without directly importing them in code). eg https://github.com/foundry-rs/foundry/issues/7025

I don't really like any of the approaches adding new configuration parameters as we are not supposed to force users to do such hacks just to make resolution working correctly. I think ideally we should be able to resolve such misses dynamically. Eg if user requests artifact that we are missing, we'll just compile it on first cheat invocation. There are also cache improvements we can do wrt this

We can consider introducing `always_compile` option to config as a temp hack but tbh I don't think it's much cleaner than just keeeping some helpers in `src/test`.

**Comment by meetmangukiya** [comment]
On the fly compilation for requested artifact within the cheatcode call is a good solution imo

**Comment by 0xGuybrush** [comment]
I'm getting the same issue (can't directly import because of pinned clashing `solc` version in pragma). 

My workaround for now (using @meetmangukiya's sample MCVE) was to use FFI. 
Rather than:

```solidity
bytes memory code = vm.getCode("test/tokens/ERC20.sol:ERC20");
```

Instead doing:

```solidity
string[] memory inputs = new string[](4);
inputs[0] = "forge";
inputs[1] = "inspect";
inputs[2] = "ERC20";
inputs[3] = "bytecode";
bytes memory code = vm.ffi(inputs);
```
which picks up the artifact correctly

**Comment by mshakeg** [comment]
this is still an issue on 1.3.1-nightly
